gateway/src/cron.rs

```diff
@@ -0,0 +1,427 @@
+use crate::StorableMsg;
+use anyhow::anyhow;
+use cid::multihash::Code;
+use cid::multihash::MultihashDigest;
+use fvm_ipld_blockstore::Blockstore;
+use fvm_ipld_encoding::to_vec;
+use fvm_ipld_encoding::tuple::{Deserialize_tuple, Serialize_tuple};
+use fvm_ipld_hamt::BytesKey;
+use fvm_shared::address::Address;
+use fvm_shared::clock::ChainEpoch;
+use ipc_sdk::ValidatorSet;
+use primitives::{TCid, THamt};
+use std::cmp::Ordering;
+use std::collections::HashSet;
+
+pub type HashOutput = Vec<u8>;
+const RATIO_NUMERATOR: u16 = 2;
+const RATIO_DENOMINATOR: u16 = 3;
+
+/// Checkpoints propagated from parent to child to signal the "final view" of the parent chain
+/// from the different validators in the subnet.
+#[derive(Clone, Debug, Serialize_tuple, Deserialize_tuple, PartialEq, Eq)]
+pub struct CronCheckpoint {
+    pub epoch: ChainEpoch,
+    pub validators: ValidatorSet,
```

Comment:
In the end you can remove validators from here, we will get the validator set from the new data structure introduced here: https://github.com/consensus-shipyard/ipc-actors/issues/68. Thanks!